// es5-ext/object/index.js
"use strict";

module.exports = {
	assign: require("./assign"),
	assignDeep: require("./assign-deep"),
	clear: require("./clear"),
	compact: require("./compact"),
	compare: require("./compare"),
	copy: require("./copy"),
	copyDeep: require("./copy-deep"),
	count: require("./count"),
	create: require("./create"),
	ensureArray: require("./ensure-array"),
	ensureFiniteNumber: require("./ensure-finite-number"),
	ensureInteger: require("./ensure-integer"),
	ensureNaturalNumber: require("./ensure-natural-number"),
	ensureNaturalNumberValue: require("./ensure-natural-number-value"),
	ensurePlainFunction: require("./ensure-plain-function"),
	ensurePromise: require("./ensure-promise"),
	ensureThenable: require("./ensure-thenable"),
	entries: require("./entries"),
	eq: require("./eq"),
	every: require("./every"),
	filter: require("./filter"),
	find: require("./find"),
	findKey: require("./find-key"),
	firstKey: require("./first-key"),
	flatten: require("./flatten"),
	forEach: require("./for-each"),
	getPropertyNames: require("./get-property-names"),
	is: require("./is"),
	isArrayLike: require("./is-array-like"),
	isCallable: require("./is-callable"),
	isCopy: require("./is-copy"),
	isCopyDeep: require("./is-copy-deep"),
	isEmpty: require("./is-empty"),
	isFiniteNumber: require("./is-finite-number"),
	isInteger: require("./is-integer"),
	isNaturalNumber: require("./is-natural-number"),
	isNaturalNumberValue: require("./is-natural-number-value"),
	isNumberValue: require("./is-number-value"),
	isObject: require("./is-object"),
	isPlainFunction: require("./is-plain-function"),
	isPlainObject: require("./is-plain-object"),
	isPromise: require("./is-promise"),
	isThenable: require("./is-thenable"),
	isValue: require("./is-value"),
	keyOf: require("./key-of"),
	keys: require("./keys"),
	map: require("./map"),
	mapKeys: require("./map-keys"),
	normalizeOptions: require("./normalize-options"),
	mixin: require("./mixin"),
	mixinPrototypes: require("./mixin-prototypes"),
	primitiveSet: require("./primitive-set"),
	safeTraverse: require("./safe-traverse"),
	serialize: require("./serialize"),
	setPrototypeOf: require("./set-prototype-of"),
	some: require("./some"),
	toArray: require("./to-array"),
	unserialize: require("./unserialize"),
	validateArrayLike: require("./validate-array-like"),
	validateArrayLikeObject: require("./validate-array-like-object"),
	validCallable: require("./valid-callable"),
	validObject: require("./valid-object"),
	validateStringifiable: require("./validate-stringifiable"),
	validateStringifiableValue: require("./validate-stringifiable-value"),
	validValue: require("./valid-value")
};
