// probot/lib/application.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var express_1 = __importDefault(require("express"));
var promise_events_1 = require("promise-events");
var context_1 = require("./context");
var github_1 = require("./github");
var logger_1 = require("./logger");
var wrap_logger_1 = require("./wrap-logger");
// Some events can't get an authenticated client (#382):
function isUnauthenticatedEvent(event) {
    return !event.payload.installation ||
        (event.name === 'installation' && event.payload.action === 'deleted');
}
/**
 * The `app` parameter available to `ApplicationFunction`s
 *
 * @property {logger} log - A logger
 */
var Application = /** @class */ (function () {
    function Application(options) {
        var opts = options || {};
        this.events = new promise_events_1.EventEmitter();
        this.log = wrap_logger_1.wrapLogger(logger_1.logger, logger_1.logger);
        this.app = opts.app;
        this.cache = opts.cache;
        this.router = opts.router || express_1.default.Router(); // you can do this?
        this.githubToken = opts.githubToken;
        if (opts.catchErrors) {
            // Deprecated since 7.2.0
            // tslint:disable-next-line:no-console
            console.warn(new Error('Propery `catchErrors` is deprecated and has no effect'));
        }
    }
    /**
     * Loads an ApplicationFunction into the current Application
     * @param appFn - Probot application function to load
     */
    Application.prototype.load = function (appFn) {
        var _this = this;
        if (Array.isArray(appFn)) {
            appFn.forEach(function (a) { return _this.load(a); });
        }
        else {
            appFn(this);
        }
        return this;
    };
    Application.prototype.receive = function (event) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                if (event.event) {
                    // tslint:disable-next-line:no-console
                    console.warn(new Error('Propery `event` is deprecated, use `name`'));
                    event = __assign({ name: event.event }, event);
                }
                return [2 /*return*/, Promise.all([
                        this.events.emit('*', event),
                        this.events.emit(event.name, event),
                        this.events.emit(event.name + "." + event.payload.action, event)
                    ])];
            });
        });
    };
    /**
     * Get an {@link http://expressjs.com|express} router that can be used to
     * expose HTTP endpoints
     *
     * ```
     * module.exports = app => {
     *   // Get an express router to expose new HTTP endpoints
     *   const route = app.route('/my-app');
     *
     *   // Use any middleware
     *   route.use(require('express').static(__dirname + '/public'));
     *
     *   // Add a new route
     *   route.get('/hello-world', (req, res) => {
     *     res.end('Hello World');
     *   });
     * };
     * ```
     *
     * @param path - the prefix for the routes
     * @returns an [express.Router](http://expressjs.com/en/4x/api.html#router)
     */
    Application.prototype.route = function (path) {
        if (path) {
            var router = express_1.default.Router();
            this.router.use(path, router);
            return router;
        }
        else {
            return this.router;
        }
    };
    /**
     * Listen for [GitHub webhooks](https://developer.github.com/webhooks/),
     * which are fired for almost every significant action that users take on
     * GitHub.
     *
     * @param event - the name of the [GitHub webhook
     * event](https://developer.github.com/webhooks/#events). Most events also
     * include an "action". For example, the * [`issues`](
     * https://developer.github.com/v3/activity/events/types/#issuesevent)
     * event has actions of `assigned`, `unassigned`, `labeled`, `unlabeled`,
     * `opened`, `edited`, `milestoned`, `demilestoned`, `closed`, and `reopened`.
     * Often, your bot will only care about one type of action, so you can append
     * it to the event name with a `.`, like `issues.closed`.
     *
     * ```js
     * app.on('push', context => {
     *   // Code was just pushed.
     * });
     *
     * app.on('issues.opened', context => {
     *   // An issue was just opened.
     * });
     * ```
     *
     * @param callback - a function to call when the
     * webhook is received.
     */
    Application.prototype.on = function (eventName, callback) {
        var _this = this;
        if (typeof eventName === 'string') {
            return this.events.on(eventName, function (event) { return __awaiter(_this, void 0, void 0, function () {
                var log, github, context, err_1;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            log = this.log.child({ name: 'event', id: event.id });
                            _a.label = 1;
                        case 1:
                            _a.trys.push([1, 4, , 5]);
                            return [4 /*yield*/, this.authenticateEvent(event, log)];
                        case 2:
                            github = _a.sent();
                            context = new context_1.Context(event, github, log);
                            return [4 /*yield*/, callback(context)];
                        case 3:
                            _a.sent();
                            return [3 /*break*/, 5];
                        case 4:
                            err_1 = _a.sent();
                            log.error({ err: err_1, event: event });
                            throw err_1;
                        case 5: return [2 /*return*/];
                    }
                });
            }); });
        }
        else {
            eventName.forEach(function (e) { return _this.on(e, callback); });
        }
    };
    /**
     * Authenticate and get a GitHub client that can be used to make API calls.
     *
     * You'll probably want to use `context.github` instead.
     *
     * **Note**: `app.auth` is asynchronous, so it needs to be prefixed with a
     * [`await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await)
     * to wait for the magic to happen.
     *
     * ```js
     *  module.exports = (app) => {
     *    app.on('issues.opened', async context => {
     *      const github = await app.auth();
     *    });
     *  };
     * ```
     *
     * @param id - ID of the installation, which can be extracted from
     * `context.payload.installation.id`. If called without this parameter, the
     * client wil authenticate [as the app](https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/about-authentication-options-for-github-apps/#authenticating-as-a-github-app)
     * instead of as a specific installation, which means it can only be used for
     * [app APIs](https://developer.github.com/v3/apps/).
     *
     * @returns An authenticated GitHub API client
     * @private
     */
    Application.prototype.auth = function (id, log) {
        if (log === void 0) { log = this.log; }
        return __awaiter(this, void 0, void 0, function () {
            var github, installationTokenTTL, res;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (process.env.GHE_HOST && /^https?:\/\//.test(process.env.GHE_HOST)) {
                            throw new Error('Your \`GHE_HOST\` environment variable should not begin with https:// or http://');
                        }
                        github = github_1.GitHubAPI({
                            baseUrl: process.env.GHE_HOST && "https://" + process.env.GHE_HOST + "/api/v3",
                            debug: process.env.LOG_LEVEL === 'trace',
                            logger: log.child({ name: 'github', installation: String(id) })
                        });
                        if (this.githubToken) {
                            github.authenticate({ type: 'app', token: this.githubToken });
                            return [2 /*return*/, github];
                        }
                        installationTokenTTL = parseInt(process.env.INSTALLATION_TOKEN_TTL || '3540', 10);
                        if (!id) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.cache.wrap("app:" + id + ":token", function () {
                                log.trace("creating token for installation");
                                github.authenticate({ type: 'app', token: _this.app() });
                                return github.apps.createInstallationToken({ installation_id: id });
                            }, { ttl: installationTokenTTL })];
                    case 1:
                        res = _a.sent();
                        github.authenticate({ type: 'token', token: res.data.token });
                        return [3 /*break*/, 3];
                    case 2:
                        github.authenticate({ type: 'app', token: this.app() });
                        _a.label = 3;
                    case 3: return [2 /*return*/, github];
                }
            });
        });
    };
    Application.prototype.authenticateEvent = function (event, log) {
        if (this.githubToken) {
            return this.auth();
        }
        if (isUnauthenticatedEvent(event)) {
            log.debug('`context.github` is unauthenticated. See https://probot.github.io/docs/github-api/#unauthenticated-events');
            return this.auth();
        }
        return this.auth(event.payload.installation.id, log);
    };
    return Application;
}());
exports.Application = Application;
